clib/sList.ml

```diff
@@ -0,0 +1,112 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+type 'a t =
+| Nil
+| Cons of 'a * 'a t
+| Default of int * 'a t
+(* Invariant: no two consecutive [Default] nodes, the integer is always > 0 *)
+
+let empty = Nil
+let cons x l = Cons (x, l)
+let defaultn n l =
+  if Int.equal n 0 then l
+  else match l with
+  | Nil | Cons _ -> Default (n, l)
+  | Default (m, l) -> Default (n + m, l)
+
+let default l = match l with
+| Nil | Cons _ -> Default (1, l)
+| Default (m, l) -> Default (m + 1, l)
+
+let cons_opt o l = match o with
+| None -> default l
+| Some x -> cons x l
+
+let is_empty = function
+| Nil -> true
+| Cons _ | Default _ -> false
+
+let view = function
+| Nil -> None
+| Cons (x, l) -> Some (Some x, l)
+| Default (1, l) -> Some (None, l)
+| Default (n, l) -> Some (None, Default (n - 1, l))
+
+let rec to_list l = match l with
+| Nil -> []
+| Cons (x, l) -> Some x :: to_list l
+| Default (n, l) ->
+  let l = to_list l in
+  let rec iterate n l = if n <= 0 then l else iterate (n - 1) (None :: l) in
+  iterate n l
+
+let of_full_list l = List.fold_right cons l empty
+
+let equal eq l1 l2 = CList.for_all2 eq (to_list l1) (to_list l2)
+
+let compare cmp l1 l2 = CList.compare cmp (to_list l1) (to_list l2)
```

Comment:
Same comment as above, shouldn't we have efficient comparisons for identity substitutions?